### PR TITLE
Pragmatical/flux2

### PR DESCRIPTION
**Pull request by pragmatical**
# Type of PR

- [X] Documentation changes
- [ ] Code changes
- [ ] Test changes
- [X] CI-CD changes
- [ ] GitHub Template changes

## PR Checklist

- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] My code follows the code style of this project.
- [X] I ran lint checks locally prior to submission.
- [X] Have you checked to ensure there aren't other open Pull Requests for the same update/change?

## Purpose of PR
This PR is to demonstrate using Flux2+Kustomize to do multi cluster/region deployments employing functionality that helps to avoid duplication of files as well as simplifies deployments.  The changes are as follows:

1. Move the Flux2 setup into its own directory called flux-init.  This was done to separate the process of setting up flux (Install flux CRD's, Set Up Git Repo, Start Flux Controllers, Apply the Kustomization that will reconcile with this directory in order to maintain flux through the git repo).  This step is different than the others as it is run using kubernetes kustomization rather than applying a flux kustomization object in the traditional apply -f manner.
2. Set up deploy directory structure to have layers as described here: [Deploy Directory Structure](https://github.com/retaildevcrews/ngsa-asb/blob/pragmatical/flux2/docs/NGSAKustomizations.md)
3. Provide examples of reuse of infrastructure code between environments by having different environments use different versions of files, as well as having environments use overlays for differences needed in a shared file.  These examples can be seen in the deploy>bootstrap-dev>cluster-baseline-settings directory.  
    1. Here we see that for the dev version of bootstrap a specific file is used that is different than the one under the bootstrap-preprod side of things.  This file is 04-aad-pod-identity.yaml.  The decision to use the entire file and not an overlay was made because the differences in each file are drastic and thus not a lot would be gained by overlays - in this case it might not even be possible to do an overlay.
    2. There is also an example of overlay in this directory with the values.kured patch that is applied in the kustomization using patchesStrategicMerge patching srategy.
4. The last thing that was done was that the bootstrap kustomization was moved into the individual cluster flux-kustomization directory and included in the <cluster-specific>-kustomization.yaml file.  This was done to be able to apply all kustomizations for the cluster (bootstrap+app setup) by applying one file.  This will ensure that the right environment is bootstrapped for the cluster.  
5.  I found out that using overlays causes our checkov scanner fails on the patch as the patch is just a subset of the full file.  This required updates to the checkov workflow to install kustomize and to run checkov with the kustomize framework. 
[Docs that state info about the functionality.](https://www.checkov.io/7.Scan%20Examples/Kustomize.html)

>**NOTE:** This PR only sets up the base>environment>cluster structure for bootstrap, will create subsequent work items to do the same for the other applications.  Interestingly enough, even in just bootstrap with this change we were able to remove ~900 lines of infrastructure code.

## Does this introduce a breaking change

- [  ] YES
- [X] NO

## Validation

- [ ] Unit tests updated and ran successfully
- [X] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #issue_number (this will automatically close the issue when the PR closes)
- References #issue_number (this references the issue but does not close with PR)

### CODE REVIEW

**Review by jofultz — APPROVED** [review]

File: apps/flux-system/gotk-repo.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  name: asb-repo-flux
+  namespace: flux-system
+spec:
+  interval: 1m0s
+  ref:
+    branch: kushal/fluxv2-kustomization
+  # secretRef: # We're accessing ngs-asb as readonly repo
+  #   name: flux-system
+  url: https://github.com/retaildevcrews/ngsa-asb
```
Comment (joaquinrz) [inline apps/flux-system/gotk-repo.yaml:14]: fix line endings

File: docs/Untitled.drawio [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mxfile><diagram id="C5RBs43oDa-KdzZeNtuy" name="Page-1">7VhZc9MwEP41Hp5gfBQnecRp0wKhU2ihw6NiK7aI7DWynINfzyqWYzsOOYD0yDQv0a43K2m/b4/YcPrx/FKQNPoEAeWGbQZzwzk3bNuybQe/lGZRaLq9bqEIBQu0UaW4Zb+oVppam7OAZg1DCcAlS5tKH5KE+rKhI0LArGk2Bt7cNSUhbSlufcLb2nsWyEjfwu5U+ivKwqjc2XJ7xZOYlMb6JllEApjVVM6F4fQFgCxW8bxPuQpeGZf794t7Ppy4lx8+Zz/JV+/j3fW314WzwSE/WV1B0ET+X9d24XpKeK7jpe8qF2UABeRJQJUT03C8SMYclxYuf1ApFxpwkktAFQgZQQgJ4UOAVNuNIZHazFIyTYJ3CliURxz8SaEaMM71Hihp+y5KmRQwWWGnHKyAUMacjCj3iD8JlwftAweBjxJIqHIVIBn0XarDXVRab8/YagwyyIVPt9jpbJFEhHSbP7ewU+er8VQjd0khplIs0EBQTiSbNslMdE6EK7sKd1xo6A+ggdOiwZDEqcp6oFnySuJqBmKyhRsKl1nEJL1NyTJAMywnTb7UeYBX9UJOskyjuAPkw0CaUiHpfGtYy6euzu2yuGlxVlUKq0z/qFYlzswjAXHWAuI7Fs+XlPyXlHT3TMmyY+3MSU2Wkhh7p6j2dAMskTUTGI8zPNg6dVYb/j2b3rbYdA0tMu3G41nT7ZHI1DmMS9aT55L7hxbhcgyDNxK4CtUq5XkYIlnQXWI4g3btiiAe5dnudtHAWDFoQGLGVbyuKJ9SyXyyoakQzkLc99xHuKnYTB7ckiUhSm4l3S3JivX3iM2m02w2K7nebcwN3aZ7rG7TaWF6g+gV0NkmX+J7Yh3fsddA6D12y+/uU6SfdQl+8I5fVtPdLd/as0xr9phvHPw0CPT0p4DeXjPlyxhwHIbZpzYIlCes8cnL+ag9CYxykSwHAcjlyySwbRKwNzWhB50EykJYA/ULTQlDMM3hKc4B6//8z5zHngPKStGEgKuY2maRYSeGwfos5ljHwwDF6k1tUQqr993OxW8=</diagram></mxfile>
```
Comment (joaquinrz) [inline docs/Untitled.drawio:2]: Plz rename this file to be more descriptive, not sure what Untitled.drawio means :(  do we need it?

**Review by joaquinrz — APPROVED** [review]
Approved with some comments

File: docs/Untitled.drawio (hunk 2 above)
Comment (pragmatical) [reply docs/Untitled.drawio:2]: checked in on accident... will remove

File: .github/workflows/checkov-k8s.yaml [hunk 3]
```diff
@@ -8,13 +8,16 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2      
+      - uses: actions/checkout@v2 
+      - name: Install kustomize 
+        run: docker pull traherom/kustomize-docker
```
[the remark this answers is not in the document]
Comment (pragmatical) [reply .github/workflows/checkov-k8s.yaml]: have implemented the requested fix

File: flux-init/init-preprod/kustomization.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 resources:
-- gotk-components.yaml
+- gotk-components-pre.yaml
 - gotk-repo.yaml
-- gotk-sync.yaml
+- got-sync.yaml
```
Comment (kforeverisback) [inline flux-init/init-preprod/kustomization.yaml:6]: Should it be `gotk-sync.yaml`?

**Review by kforeverisback — COMMENTED** [review]
Couple of comments:
- There are two folders under flux-init: init-pre, init-dev. Do we need two folders for Flux since the only change is the repo address and path, we should be able to use one folder and kustomize
- In the diagram under `docs/NgsaKustomizations.md` the directory `apps-base`, `apps-dev` and `apps-pre` are listed but I don't see them in the git repo

File: .github/workflows/checkov-k8s.yaml [hunk 5]
```diff
@@ -8,13 +8,16 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2      
+      - uses: actions/checkout@v2 
+      - name: Install kustomize 
+        run: curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | bash  
+      - run: kustomize version
```
[the remark this answers is not in the document]
Comment (pragmatical) [reply .github/workflows/checkov-k8s.yaml:14]: removing this check

**Review by kforeverisback — APPROVED** [review]
LGTM ⚡ 

We'll focus on app-base and flux-init conglomeration on another task based on necessity.

**Pull request merged**